src/System.IO.Pipes/src/System/IO/Pipes/PipeStreamCompletionSource.cs

```diff
@@ -0,0 +1,225 @@
+#if DEBUG
+            Debug.Assert(!_cancellationHasBeenRegistered, "Cannot register for cancellation twice");
+            _cancellationHasBeenRegistered = true;
+#endif
+
+            // Quick check to make sure that the cancellation token supports cancellation, and that the IO hasn't completed
+            if (_cancellationToken.CanBeCanceled && Overlapped != null)
+            {
+                // Register the cancellation only if the IO hasn't completed
+                int state = Interlocked.CompareExchange(ref _state, RegisteringCancellation, NoResult);
+                if (state == NoResult)
+                {
+                    // Register the cancellation
+                    _cancellationRegistration = _cancellationToken.Register(thisRef => ((PipeStreamCompletionSource)thisRef).Cancel(), this);
+
+                    // Grab the state for case if IO completed while we were setting the registration.
+                    state = Interlocked.Exchange(ref _state, NoResult);
+                }
+                else if (state != CompletedCallback)
+                {
+                    // IO already completed and we have grabbed result state.
+                    // Set NoResult to prevent invocation of CompleteCallback(result state) from AsyncCallback(...)
+                    state = Interlocked.Exchange(ref _state, NoResult);
+                }
+
+                // If we have the result state of completed IO call CompleteCallback(result).
+                // Otherwise IO not completed.
+                if (state == ResultSuccess || state == ResultError)
+                {
+                    CompleteCallback(state);
+                }
+            }
+        }
+
+        internal void ReleaseResources()
+        {
+            _cancellationRegistration.Dispose();
+
+            // NOTE: The cancellation must *NOT* be running at this point, or it may observe freed memory
+            // (this is why we disposed the registration above)
+            if (Overlapped != null)
+            {
+                _threadPoolBinding.FreeNativeOverlapped(Overlapped);
+                _overlapped = null;
+            }
+        }
+
+        internal void SetCompletedSynchronously(int numBytes = default(int))
+        {
+            if (!_isWrite)
+            {
+                _pipeStream.UpdateMessageCompletion(_isMessageComplete);
+            }
+
+            TrySetResult(numBytes);
+        }
+
+        private void AsyncCallback(uint errorCode, uint numBytes)
+        {
+            _numBytes = (int)numBytes;
+
+            // Allow async read to finish
+            if (!_isWrite)
+            {
+                switch (errorCode)
+                {
+                    case Interop.mincore.Errors.ERROR_BROKEN_PIPE:
+                    case Interop.mincore.Errors.ERROR_PIPE_NOT_CONNECTED:
+                    case Interop.mincore.Errors.ERROR_NO_DATA:
+                        errorCode = 0;
+                        numBytes = 0;
```

Comment:
Value of `numBytes` never used after assignment. Original callback (`AsyncPSCallback`) does the same. How should I fix this?